From 38dc120ef7e58f72bbabe0cc2b7e1949f5e3187d Mon Sep 17 00:00:00 2001
From: Oliver Beard <olib141@outlook.com>
Date: Fri, 16 Jan 2026 16:59:42 +0000
Subject: [PATCH] FROMLIST: frontend: Use GreeterFader & make Show clock option
 consistent with screen locker

WallpaperFader has been changed to no longer depend on greeter components, so it can now be used directly. The greeter can use GreeterFader as well for consistency.

The 'Show clock' option has also been made consistent with the screen locker as it ties into the GreeterFader functionality. We previously did not support only showing the clock on the login prompt.

Link: https://invent.kde.org/plasma/plasma-login-manager/-/merge_requests/85
Signed-off-by: Mingcong Bai <jeffbai@aosc.io>
---
 CMakeLists.txt                                |  2 +-
 src/frontend/greeter/qml/Main.qml             | 32 +++-----
 src/frontend/kcm/ui/Appearance.qml            | 34 +++++++--
 .../settings/plasmaloginsettingsbase.kcfg     |  7 +-
 .../settings/plasmaloginsettingsdefaults.cpp  |  6 +-
 .../settings/plasmaloginsettingsdefaults.h    |  6 +-
 src/frontend/wallpaper/CMakeLists.txt         | 12 ---
 src/frontend/wallpaper/WallpaperFader.frag    | 22 ------
 src/frontend/wallpaper/WallpaperFader.qml     | 73 -------------------
 src/frontend/wallpaper/main.qml               |  4 +-
 10 files changed, 55 insertions(+), 143 deletions(-)
 delete mode 100644 src/frontend/wallpaper/WallpaperFader.frag
 delete mode 100644 src/frontend/wallpaper/WallpaperFader.qml

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 3f401f8..5a38932 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -37,7 +37,7 @@ include(ECMQtDeclareLoggingCategory)
 include(ECMConfiguredInstall)
 include(ECMQmlModule)
 
-find_package(Qt6 CONFIG REQUIRED Core DBus Gui Qml Quick LinguistTools Test QuickTest ShaderTools)
+find_package(Qt6 CONFIG REQUIRED Core DBus Gui Qml Quick LinguistTools Test QuickTest)
 find_package(KF6 ${KF6_MIN_VERSION} REQUIRED COMPONENTS Config Package WindowSystem I18n DBusAddons KCMUtils Auth KIO CoreAddons)
 find_package(PlasmaQuick ${PROJECT_DEP_VERSION} REQUIRED)
 find_package(LayerShellQt ${PROJECT_DEP_VERSION} REQUIRED)
diff --git a/src/frontend/greeter/qml/Main.qml b/src/frontend/greeter/qml/Main.qml
index eb5b499..26fe1ee 100644
--- a/src/frontend/greeter/qml/Main.qml
+++ b/src/frontend/greeter/qml/Main.qml
@@ -76,29 +76,30 @@ Item {
         onPressed: PlasmaLogin.GreeterState.activateWindow(Window.window);
         onPositionChanged: PlasmaLogin.GreeterState.activateWindow(Window.window);
 
+        BreezeComponents.GreeterFader {
+            state: loginScreenRoot.uiVisible ? "on" : "off"
+            mainStack: mainStack
+            footer: footer
+            clock: clock
+            alwaysShowClock: Settings.showClock === 0
+        }
+
         DropShadow {
             id: clockShadow
             anchors.fill: clock
             source: clock
-            visible: !softwareRendering && clock.visible
+            visible: !softwareRendering && Settings.alwaysShowClock === 0
             radius: 7
             verticalOffset: 0.8
             samples: 15
             spread: 0.2
             color: Qt.rgba(0, 0, 0, 0.7)
-            opacity: loginScreenRoot.uiVisible ? 0 : 1
-            Behavior on opacity {
-                OpacityAnimator {
-                    duration: Kirigami.Units.veryLongDuration * 2
-                    easing.type: Easing.InOutQuad
-                }
-            }
         }
 
         BreezeComponents.Clock {
             id: clock
             property Item shadow: clockShadow
-            visible: y > 0 && Settings.showClock
+            visible: y > 0 && Settings.alwaysShowClock !== 2
             anchors.horizontalCenter: parent.horizontalCenter
             y: (userListComponent.userList.y + mainStack.y)/2 - height/2
             Layout.alignment: Qt.AlignBaseline
@@ -115,12 +116,7 @@ Item {
 
             focus: true
 
-            opacity: loginScreenRoot.uiVisible ? 1 : 0
-            Behavior on opacity {
-                OpacityAnimator {
-                    duration: Kirigami.Units.longDuration
-                }
-            }
+            visible: opacity > 0
 
             Connections {
                 target: PlasmaLogin.GreeterState
@@ -361,12 +357,6 @@ Item {
 
             spacing: Kirigami.Units.smallSpacing
 
-            Behavior on opacity {
-                OpacityAnimator {
-                    duration: Kirigami.Units.longDuration
-                }
-            }
-
             /* Virtual keyboard btn */
 
             KeyboardButton {
diff --git a/src/frontend/kcm/ui/Appearance.qml b/src/frontend/kcm/ui/Appearance.qml
index 8c8c904..bdf9c51 100644
--- a/src/frontend/kcm/ui/Appearance.qml
+++ b/src/frontend/kcm/ui/Appearance.qml
@@ -31,12 +31,36 @@ Kirigami.Page {
         Kirigami.FormLayout {
             id: parentLayout // Don't change needed for correct alignment with wallpaper config
 
-            QQC2.CheckBox {
-                Kirigami.FormData.label: i18n("Clock")
-                text: i18n("Show clock")
+            QQC2.RadioButton {
+                Kirigami.FormData.label: i18nc("@title:group 'Show clock: Always/On login prompt/Never'", "Show clock:")
+                text: i18nc("@option:radio, 'Show clock: Always'", "Always")
 
-                checked: kcm.settings.showClock
-                onToggled: kcm.settings.showClock = checked
+                checked: kcm.settings.showClock === 0
+                onToggled: kcm.settings.showClock = 0
+
+                KCM.SettingStateBinding {
+                    configObject: kcm.settings
+                    settingName: "ShowClock"
+                }
+            }
+
+            QQC2.RadioButton {
+                text: i18nc("@option:radio, 'Show clock: On login prompt'", "On login prompt")
+
+                checked: kcm.settings.showClock === 1
+                onToggled: kcm.settings.showClock = 1
+
+                KCM.SettingStateBinding {
+                    configObject: kcm.settings
+                    settingName: "ShowClock"
+                }
+            }
+
+            QQC2.RadioButton {
+                text: i18nc("@option:radio, 'Show clock: Never'", "Never")
+
+                checked: kcm.settings.showClock === 2
+                onToggled: kcm.settings.showClock = 2
 
                 KCM.SettingStateBinding {
                     configObject: kcm.settings
diff --git a/src/frontend/settings/plasmaloginsettingsbase.kcfg b/src/frontend/settings/plasmaloginsettingsbase.kcfg
index fbe19ad..51cd70f 100644
--- a/src/frontend/settings/plasmaloginsettingsbase.kcfg
+++ b/src/frontend/settings/plasmaloginsettingsbase.kcfg
@@ -22,7 +22,12 @@
     <entry name="PreselectedSession" type="String">
       <default code="true">defaultPreselectedSession()</default>
     </entry>
-    <entry name="ShowClock" type="Bool">
+    <entry name="ShowClock" type="Enum">
+      <choices>
+        <choice name="Always"/>
+        <choice name="OnLoginPrompt"/>
+        <choice name="Never"/>
+      </choices>
       <default code="true">defaultShowClock()</default>
     </entry>
     <entry name="WallpaperPluginId" type="String">
diff --git a/src/frontend/settings/plasmaloginsettingsdefaults.cpp b/src/frontend/settings/plasmaloginsettingsdefaults.cpp
index 47c58a8..d36f280 100644
--- a/src/frontend/settings/plasmaloginsettingsdefaults.cpp
+++ b/src/frontend/settings/plasmaloginsettingsdefaults.cpp
@@ -13,7 +13,7 @@ QString PlasmaLoginSettingsDefaults::s_defaultSession;
 bool PlasmaLoginSettingsDefaults::s_defaultRelogin;
 QString PlasmaLoginSettingsDefaults::s_defaultPreselectedUser;
 QString PlasmaLoginSettingsDefaults::s_defaultPreselectedSession;
-bool PlasmaLoginSettingsDefaults::s_defaultShowClock;
+int PlasmaLoginSettingsDefaults::s_defaultShowClock;
 QString PlasmaLoginSettingsDefaults::s_defaultWallpaperPluginId;
 
 PlasmaLoginSettingsDefaults::PlasmaLoginSettingsDefaults(KSharedConfigPtr config, QObject *parent)
@@ -25,7 +25,7 @@ PlasmaLoginSettingsDefaults::PlasmaLoginSettingsDefaults(KSharedConfigPtr config
     s_defaultRelogin = defaultConfig->group(QStringLiteral("AutoLogin")).readEntry("Relogin", false);
     s_defaultPreselectedUser = defaultConfig->group(QStringLiteral("Greeter")).readEntry("PreselectedUser", "");
     s_defaultPreselectedSession = defaultConfig->group(QStringLiteral("Greeter")).readEntry("PreselectedSession", "");
-    s_defaultShowClock = defaultConfig->group(QStringLiteral("Greeter")).readEntry("ShowClock", true);
+    s_defaultShowClock = defaultConfig->group(QStringLiteral("Greeter")).readEntry("ShowClock", 0);
     s_defaultWallpaperPluginId = defaultConfig->group(QStringLiteral("Greeter")).readEntry("WallpaperPluginId", "org.kde.image");
 }
 
@@ -54,7 +54,7 @@ QString PlasmaLoginSettingsDefaults::defaultPreselectedSession()
     return s_defaultPreselectedSession;
 }
 
-bool PlasmaLoginSettingsDefaults::defaultShowClock()
+int PlasmaLoginSettingsDefaults::defaultShowClock()
 {
     return s_defaultShowClock;
 }
diff --git a/src/frontend/settings/plasmaloginsettingsdefaults.h b/src/frontend/settings/plasmaloginsettingsdefaults.h
index d320c68..22ed4c2 100644
--- a/src/frontend/settings/plasmaloginsettingsdefaults.h
+++ b/src/frontend/settings/plasmaloginsettingsdefaults.h
@@ -17,7 +17,7 @@ class PlasmaLoginSettingsDefaults : public KConfigSkeleton
     Q_PROPERTY(bool defaultRelogin READ defaultRelogin CONSTANT)
     Q_PROPERTY(QString defaultPreselectedUser READ defaultPreselectedUser CONSTANT)
     Q_PROPERTY(QString defaultPreselectedSession READ defaultPreselectedSession CONSTANT)
-    Q_PROPERTY(bool defaultShowClock READ defaultShowClock CONSTANT)
+    Q_PROPERTY(int defaultShowClock READ defaultShowClock CONSTANT)
     Q_PROPERTY(QString defaultWallpaperPluginId READ defaultWallpaperPluginId CONSTANT)
 
 public:
@@ -28,7 +28,7 @@ public:
     static bool defaultRelogin();
     static QString defaultPreselectedUser();
     static QString defaultPreselectedSession();
-    static bool defaultShowClock();
+    static int defaultShowClock();
     static QString defaultWallpaperPluginId();
 
 private:
@@ -37,6 +37,6 @@ private:
     static bool s_defaultRelogin;
     static QString s_defaultPreselectedUser;
     static QString s_defaultPreselectedSession;
-    static bool s_defaultShowClock;
+    static int s_defaultShowClock;
     static QString s_defaultWallpaperPluginId;
 };
diff --git a/src/frontend/wallpaper/CMakeLists.txt b/src/frontend/wallpaper/CMakeLists.txt
index d122038..beee78d 100644
--- a/src/frontend/wallpaper/CMakeLists.txt
+++ b/src/frontend/wallpaper/CMakeLists.txt
@@ -11,7 +11,6 @@ target_sources(plasma-login-wallpaper PRIVATE
 ecm_target_qml_sources(plasma-login-wallpaper PRIVATE
     SOURCES
     main.qml
-    WallpaperFader.qml
 )
 
 target_link_libraries(plasma-login-wallpaper PRIVATE
@@ -26,17 +25,6 @@ target_link_libraries(plasma-login-wallpaper PRIVATE
     LayerShellQt::Interface
 )
 
-qt_add_shaders(plasma-login-wallpaper "wallpaper_shaders"
-    PRECOMPILE
-    BATCHABLE
-    OPTIMIZED
-    PREFIX "/qt/qml/org/kde/plasma/login/wallpaper/shaders"
-    FILES
-        WallpaperFader.frag
-    OUTPUTS
-        WallpaperFader.frag.qsb
-)
-
 ecm_finalize_qml_module(plasma-login-wallpaper)
 
 install(TARGETS plasma-login-wallpaper DESTINATION ${KDE_INSTALL_TARGETS_DEFAULT_ARGS})
diff --git a/src/frontend/wallpaper/WallpaperFader.frag b/src/frontend/wallpaper/WallpaperFader.frag
deleted file mode 100644
index 736860a..0000000
--- a/src/frontend/wallpaper/WallpaperFader.frag
+++ /dev/null
@@ -1,22 +0,0 @@
-/*
-    SPDX-FileCopyrightText: 2014 Aleix Pol Gonzalez <aleixpol@blue-systems.com>
-    SPDX-FileCopyrightText: 2023 David Redondo <kde@david-redondo.de>
-
-    SPDX-License-Identifier: GPL-2.0-or-later
-*/
-#version 440
-
-layout(location = 0) in vec2 qt_TexCoord0;
-layout(std140, binding = 0) uniform buf {
-    mat4 qt_Matrix;
-    float qt_Opacity;
-    mat4 colorMatrix;
-};
-layout(binding = 1) uniform mediump sampler2D source;
-layout(location = 0) out vec4 fragColor;
-
-void main(void)
-{
-    mediump vec4 tex = texture(source, qt_TexCoord0);
-    fragColor = tex * colorMatrix * qt_Opacity;
-}
diff --git a/src/frontend/wallpaper/WallpaperFader.qml b/src/frontend/wallpaper/WallpaperFader.qml
deleted file mode 100644
index 23909e7..0000000
--- a/src/frontend/wallpaper/WallpaperFader.qml
+++ /dev/null
@@ -1,73 +0,0 @@
-/*
-    SPDX-FileCopyrightText: 2014 Aleix Pol Gonzalez <aleixpol@blue-systems.com>
-
-    SPDX-License-Identifier: GPL-2.0-or-later
-*/
-
-import QtQuick
-import Qt5Compat.GraphicalEffects
-
-import org.kde.kirigami as Kirigami
-
-Item {
-    id: wallpaperFader
-
-    property alias source: wallpaperBlur.source
-    property real factor: 0
-    readonly property bool lightColorScheme: Math.max(Kirigami.Theme.backgroundColor.r, Kirigami.Theme.backgroundColor.g, Kirigami.Theme.backgroundColor.b) > 0.5
-
-    Behavior on factor {
-        NumberAnimation {
-            target: wallpaperFader
-            property: "factor"
-            duration: Kirigami.Units.veryLongDuration * 2
-            easing.type: Easing.InOutQuad
-        }
-    }
-
-    FastBlur {
-        id: wallpaperBlur
-        anchors.fill: parent
-        radius: 50 * wallpaperFader.factor
-    }
-
-    ShaderEffect {
-        id: wallpaperShader
-        anchors.fill: parent
-
-        supportsAtlasTextures: true
-
-        property var source: ShaderEffectSource {
-            sourceItem: wallpaperBlur
-            live: true
-            hideSource: true
-            textureMirroring: ShaderEffectSource.NoMirroring
-        }
-
-        readonly property real contrast: 0.8 * wallpaperFader.factor + (1 - wallpaperFader.factor)
-        readonly property real saturation: 1.5 * wallpaperFader.factor + (1 - wallpaperFader.factor)
-        readonly property real intensity: (wallpaperFader.lightColorScheme ? 1.6 : 0.7) * wallpaperFader.factor + (1 - wallpaperFader.factor)
-
-        readonly property real transl: (1.0 - contrast) / 2.0;
-        readonly property real rval: (1.0 - saturation) * 0.2126;
-        readonly property real gval: (1.0 - saturation) * 0.7152;
-        readonly property real bval: (1.0 - saturation) * 0.0722;
-
-        property var colorMatrix: Qt.matrix4x4(
-            contrast, 0,        0,        0.0,
-            0,        contrast, 0,        0.0,
-            0,        0,        contrast, 0.0,
-            transl,   transl,   transl,   1.0).times(Qt.matrix4x4(
-                rval + saturation, rval,     rval,     0.0,
-                gval,     gval + saturation, gval,     0.0,
-                bval,     bval,     bval + saturation, 0.0,
-                0,        0,        0,        1.0)).times(Qt.matrix4x4(
-                    intensity, 0,         0,         0,
-                    0,         intensity, 0,         0,
-                    0,         0,         intensity, 0,
-                    0,         0,         0,         1
-                ));
-
-        fragmentShader: "qrc:/qt/qml/org/kde/plasma/login/wallpaper/shaders/WallpaperFader.frag.qsb"
-    }
-}
diff --git a/src/frontend/wallpaper/main.qml b/src/frontend/wallpaper/main.qml
index 2b807d7..7217c76 100644
--- a/src/frontend/wallpaper/main.qml
+++ b/src/frontend/wallpaper/main.qml
@@ -10,7 +10,7 @@ import QtQuick.Window
 
 import org.kde.kirigami as Kirigami
 
-import org.kde.plasma.login.wallpaper as PlasmaLoginWallpaper
+import org.kde.breeze.components as BreezeComponents
 
 Item {
     id: main
@@ -26,7 +26,7 @@ Item {
         anchors.fill: parent
     }
 
-    PlasmaLoginWallpaper.WallpaperFader {
+    BreezeComponents.WallpaperFader {
         anchors.fill: parent
         factor: Window.window?.blur ? 1 : 0
         source: wallpaperPlaceholder.children[0]
-- 
2.52.0

